beku.nix

```
@@ -1,11 +1,19 @@
-{ python3 }:
-with python3;
+{ python3, pkgs }:
```

```suggestion
{ python3, runCommand, lib }:
```

Generally, the convention for `pkgs.callPackage` is to take every dependency as a separate argument (just like we already take `python3` instead of referring to `pkgs.python3`).

---

Yesn't. `pkgs.callPackage ./foo.nix bar` calls `foo.nix` with the single argument `pkgs // bar` (`pkgs` overridden with any values from `bar`). This is then destructed, as you say. (There is some additional magic going on so that only the arguments actually used are passed, normally you need `, ...` to have the destructure ignore everything else.)

Because of the way that Nixpkgs is set up, `pkgs.pkgs = pkgs`, so `pkgs.pkgs.pkgs.pkgs.pkgs.pkgs` will still just return `pkgs`. Normally if you wanted to destructure while also keeping the original object (attrset) you'd use `{ foo, bar, ... } @ baz:`.

---

Oh cool, I have see that (`@` notation) before, just not at the top, so it seemed completely unrelated in my (broken) mental model. Your explanation helped demystify a whole bunch of stuff, so thanks.

---

Done in 661fa0b
I test it locally before merging.

---

More fixes in 43bb39cf8f57d33c617fa7fd6dc36b337be7a8ef